[EDR Workflows] Add crowdstrike agent status

tomsonpl opened the pull request:
### Overview:
_This PR represents a smaller component of a larger functionality aimed at supporting bi-directional actions against Crowdstrike._

It introduces functionality to retrieve **Crowdstrike** agent statuses, specifically both `containment` and `online` status.

### UI Changes:
<img width="752" alt="Zrzut ekranu 2024-05-8 o 17 34 18" src="https://github.com/elastic/kibana/assets/16632552/a07e38aa-1709-4733-bcf4-d4c4a57ea68b">

### Key Features:
- Enhances system monitoring capabilities by providing real-time insights into Crowdstrike agent statuses.

### Changes Implemented:
- Calculate and render agent healthy badge + isolation badge
- Registered new SubAction to getAgentOnlineStatus
- Get AgentDetails from Host event logs
- Get AgentOnlineStatus from API
- Fixed schemas in connector side (the subaction hasn't been previously used yet) 
- Prepared Initial Version (happy path) of external edr crowdstrike server data

**When agentStatusClientEnabled is true:**
- Utilizes the Crowdstrike client's status service to obtain agent details from ES logs.
- Fetches online status via API.

**When agentStatusClientEnabled is false:**
- It does not return agent status

### Additional Notes:
- Apparently Crowdstrike's host API has 2 endpoints we can make use of.
1. First returns a wealth of  data regarding the host, including a `status` which can held one of four values: `['normal', 'contained', 'containment_pending', 'lift_containment_pending']`. 
2. Second is used to get `online_status` as `state` with values: `['online', 'offline', 'unknown']`
- Thanks to @paul-tavares and his on-week [project ](https://github.com/elastic/kibana/pull/182990) the API is mocked with a fake server.

### Impact:
- Minimal impact expected since Crowdstrike bi-direction actions are hidden behind multiple feature flags.
- SentinelOne agent status functionality should remain unchanged

### Steps to reproduce:
1. Enable `crowdstrikeConnectorOn`, `responseActionsCrowdstrikeManualHostIsolationEnabled` and `agentStatusClientEnabled` feature flags
2. Add Crowdstrike integration
3. Run Api Emulator: https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/scripts/endpoint/api_emulator/README.md#L8 
4. Add Crowdstrike connector - passing crowdstrike url from emulator as in this example: `http://MacBook-Pro-Tomasz.local:51462/crowdstrike` 
5. Use these snippets in the console to create events and an alert: 
- CS Alert event: https://p.elstc.co/paste/7o1oWTqe#BGuXc2ieVse8GWyQQae9j7tLljFbaasgIJ-JL1udKTW
- CS Host event: https://p.elstc.co/paste/Lb4n8x49#+bQjKK3luVZfX085CG8bbJNQzAKqVGGXUJzRkdaPxLB
- Rule Alert: https://p.elstc.co/paste/VMxsEh6A#557MZZ4Gt4gFCCd7TYE+kBS9xe1hM+rfUU1sGxOR5L+
6. Go to Alert's flyout and see that the status shows `Healthy`

tomsonpl committed:
add crowdstrike status - hardcoded

tomsonpl committed:
Merge branch 'main' into crowdstrike-agent-status

# Conflicts:
#	x-pack/plugins/security_solution/public/common/utils/crowdstrike_alert_check.ts
#	x-pack/plugins/security_solution/public/detections/components/host_isolation/index.tsx
#	x-pack/plugins/security_solution/public/detections/components/host_isolation/use_host_isolation_action.tsx
#	x-pack/plugins/security_solution/public/management/components/endpoint_responder/components/offline_callout.test.tsx

tomsonpl commented:
/ci

tomsonpl committed:
Merge branch 'main' into crowdstrike-agent-status

tomsonpl committed:
make use of getAgentStatus

tomsonpl commented:
/ci

tomsonpl committed:
offline callout

tomsonpl committed:
add api call to fetch status online

tomsonpl committed:
api call

tomsonpl commented:
/ci

tomsonpl committed:
fix

tomsonpl commented:
/ci

tomsonpl committed:
fix

tomsonpl committed:
fix

tomsonpl commented:
/ci

tomsonpl committed:
missing registered action

tomsonpl committed:
fix

tomsonpl commented:
/ci

tomsonpl committed:
fix

tomsonpl commented:
/ci

tomsonpl committed:
fix

tomsonpl commented:
/ci

elasticmachine commented:
Pinging @elastic/security-defend-workflows (Team:Defend Workflows)

tomsonpl commented:
This PR is missing some unit tests, but it would be awesome if I could get an implementation initial review :) 

The Crowdstrike functionality doesn't get released (still is behind a feature flag set to false) until we get to test it heavily. 

Thank you!

tomsonpl committed:
tests

tomsonpl committed:
Merge branch 'main' into crowdstrike-agent-status

rylnd commented on x-pack/plugins/security_solution/public/detections/components/host_isolation/crowdstrike_agent_status.tsx:
```diff
@@ -0,0 +1,90 @@
+    const { data, isLoading, isFetched } = useAgentStatus([agentId], 'crowdstrike', {
+      enabled: crowdstrikeManualHostActionsEnabled,
+    });
+    const agentStatus = data?.[`${agentId}`];
```
Nit: `agentId` is declared as a string, so this templating/wrapping/casting is unnecessary:

```suggestion
    const agentStatus = data?.[agentId];
```

rylnd commented on x-pack/plugins/security_solution/public/detections/components/host_isolation/use_external_host_isolation.tsx:
```diff
@@ -22,9 +22,9 @@ interface ErrorType {
 }
 
 // TODO: 8.15: Remove `useGetSentinelOneAgentStatus` function when `agentStatusClientEnabled` is enabled/removed
```
This comment should be updated or removed.

tomsonpl replied:
@rylnd Thank you for the review and suggestions, you're absolutely right, I'll make the changes :+1:

rylnd reviewed (approved):
Functional changes LGTM. I called out one comment that's been orphaned by a rename; but approving otherwise.

tomsonpl committed:
Merge branch 'main' into crowdstrike-agent-status

pmuellr commented:
I assume all the connector execution is happening via calls to the action executor directly, and not "queued up" as alerting actions?

The distinction is that we have backwards compatibilty issues when connector data "shapes" (params mainly) change between releases, in case new "shapes" get queued in response to alerts, but then we have to roll back Kibana to the previous release, which may have issues with the new "shapes".  Direct invocation doesn't have as many issues like this.

tomsonpl commented:
@pmuellr You're right, we're using the `execute` functionality, it's not being automatically triggered in rules.

tomsonpl committed:
initial version of crowdstrike mocked responses

tomsonpl committed:
Merge branch 'crowdstrike-agent-status' of github.com:tomsonpl/kibana into crowdstrike-agent-status

tomsonpl committed:
fix

tomsonpl committed:
fix

pmuellr reviewed (approved):
ResponseOps changes LGTM

tomsonpl committed:
fix type

tomsonpl committed:
Merge branch 'main' into crowdstrike-agent-status

tomsonpl committed:
Merge branch 'main' into crowdstrike-agent-status

tomsonpl committed:
merge main

tomsonpl commented:
/ci

tomsonpl committed:
adjust to new agent status

tomsonpl commented:
/ci

tomsonpl committed:
fix

tomsonpl commented:
/ci

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

tomsonpl commented:
/ci

tomsonpl committed:
fix

tomsonpl commented:
/ci

tomsonpl committed:
Merge branch 'crowdstrike-agent-status' of github.com:tomsonpl/kibana into crowdstrike-agent-status

tomsonpl commented:
/ci

PhilippeOberti commented on x-pack/plugins/security_solution/public/flyout/document_details/shared/hooks/use_highlighted_fields.ts:
```diff
@@ -111,6 +115,13 @@ export const useHighlightedFields = ({
       return acc;
     }
 
+    if (
+      field.id === CROWDSTRIKE_AGENT_ID_FIELD &&
+      !isAlertFromCrowdstrikeEvent({ data: dataFormattedForFieldBrowser })
+    ) {
+      return acc;
+    }
+
```
could you add a one line documentation here like it's not for the other if/else conditions above? Also I feel like adding a unit test for this would be nice

PhilippeOberti reviewed (commented):
LGTM for the Threat hunting Investigations team, just a missing comment and unit test

paul-tavares commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/lib/types.ts:
```diff
@@ -5,8 +5,20 @@
+  logger: Logger;
+}
+
+export type AgentType = 'endpoint' | 'sentinel_one' | 'crowdstrike';
```
Please use the existing type `ResponseActionAgentType` defined here: https://github.com/elastic/kibana/blob/cd3e684a895427652c9bea635c635905064e0b4c/x-pack/plugins/security_solution/common/endpoint/service/response_actions/constants.ts#L16

tomsonpl replied:
Right! thank you! :)

paul-tavares reviewed (changes requested):
I only got up to the changes to Stack Connectors and will pick this review again tomorrow. 

The main issue I'm confused about is: why do you have conditional logic for retrieving agent statuses for Crowdstrike?  It should only use the new service/api that @ashokaditya  created.

Lets discuss further tomorrow if you feel that should not be the case.

PhilippeOberti commented on x-pack/plugins/security_solution/public/common/components/event_details/get_alert_summary_rows.tsx:
```diff
@@ -297,15 +307,14 @@ export function getEventCategoriesFromData(data: TimelineEventsDetailsItem[]): E
 }
 
-export const getSummaryRows = ({
+export const useSummaryRows = ({
```
I thought about that as well, but didn't comment as this code is just changed, not really new... Always a good idea to add tests though, thanks for pointing it out!

tomsonpl committed:
agenttype

tomsonpl committed:
getAgentStatusFromConnectorAction

ashokaditya reviewed (changes requested):
I did a review of the agent client side only and we need some changes there. I'll do another round once you've the changes in.

Thanks for sticking with it. 🔥  I know it is not easy, so let me know if you want help with any of it.

tomsonpl committed:
Update x-pack/plugins/security_solution/public/detections/components/host_isolation/use_host_isolation_action.tsx

Co-authored-by: Paul Tavares <56442535+paul-tavares@users.noreply.github.com>

tomsonpl commented on x-pack/plugins/security_solution/public/detections/components/host_isolation/use_host_isolation_action.test.tsx:
```diff
@@ -116,8 +129,8 @@ describe('useHostIsolationAction', () => {
 
-      expect(hook).toHaveBeenCalledWith([''], 'sentinel_one', {
-        enabled: false,
+      expect(hook).toHaveBeenCalledWith(['expectedCrowdstrikeAgentId'], 'crowdstrike', {
```
These changes do not drop SentinelOne tests. Just the values were temporarily defaulted to SentinelOne before (introduced in Action Client PR). Now correctly expecting crowdstrike agentType, and `expectedCrowdstrikeAgentId` is just a value for agent.id

paul-tavares replied:
got it. thank you... So where are the checks that sentinelone one is also working as expected? Because your changes seem to be checking for crowdstrike only

tomsonpl replied:
The same file, there are 2 crowdstrike, and 2 sentinel_one tests.

tomsonpl commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/lib/base_agent_status_client.ts:
```diff
@@ -17,6 +18,7 @@ export interface AgentStatusClientOptions {
   endpointService: EndpointAppContextService;
   esClient: ElasticsearchClient;
   soClient: SavedObjectsClientContract;
+  connectorActionsClient: ActionsClient;
```
What do you mean? I need to execute connectors action (make API call) to get the data, how do you suggest I go about it without the connectorActionsClient ?

ashokaditya replied:
We're directly querying the agent status from the agent index in Kibana. Why do we need a connector client? I don't see what do we need the connector for unless you want to make an API call for creating actions on the agent.

tomsonpl replied:
As you've probably seen in the code - we do make an API call there. Crowdstrike host logs index doesn't contain all data we need, thus - second API call.

ashokaditya replied:
In that case I'd suggest you update it to allow for `undefined`. We don't need it for sentinel agent status so no need to make it a mandatory value.

tomsonpl commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/sentinel_one/agent_status.ts:
```diff
@@ -5,31 +5,22 @@
-import { HostStatus } from '../../../../common/endpoint/types';
-import { CustomHttpRequestError } from '../../../utils/custom_http_request_error';
-
-export interface GetAgentStatusOptions {
-  // NOTE: only sentinel_one currently supported
-  agentType: ResponseActionAgentType;
-  agentIds: string[];
-  connectorActionsClient: ActionsClient;
-  logger: Logger;
-}
```
This type is now used in Crowdstrike too.

tomsonpl replied:
Brought back to previous state 👍

tomsonpl commented on x-pack/plugins/security_solution/public/detections/components/host_isolation/crowdstrike_agent_status.tsx:
```diff
@@ -0,0 +1,90 @@
+  }
+`;
+
+export const CrowdstrikeAgentStatus = React.memo(
```
This file doesn't exist anymore 👍

tomsonpl commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/sentinel_one/agent_status.ts:
```diff
@@ -5,31 +5,22 @@
  * 2.0.
  */
 
-import type { ActionsClient } from '@kbn/actions-plugin/server';
```
But this SentinelOne specific logic. We have another file for crowdstrike.

ashokaditya replied:
Right. In that case I'd suggest you prefix the file names with agent type names. Also move them out to `agent/` or `agent/connectors/` folder. That way we don't mix connector logic with the new logic within `clients` folder

tomsonpl commented on x-pack/plugins/security_solution/public/flyout/document_details/right/components/highlighted_fields_cell.tsx:
```diff
@@ -96,6 +98,12 @@ export const HighlightedFieldsCell: VFC<HighlightedFieldsCellProps> = ({
           >
             {field === HOST_NAME_FIELD_NAME || field === USER_NAME_FIELD_NAME ? (
               <LinkFieldCell value={value} />
+            ) : field === AGENT_STATUS_FIELD_NAME &&
```
Not used anymore

tomsonpl reviewed (commented):
@ashokaditya @paul-tavares Thank you both for the reviews 👍. I've addressed some of the questions and will make the necessary adjustments to the code as soon as possible.

Before we proceed, I think it's important to align on the approach for fetching agent status, specifically regarding the old and new methods. My understanding is that the new method is activated when the `agentStatusClientEnabled` FF is enabled. Since this feature flag can be toggled by the user through configuration at any time, I believe we should continue supporting both methods for now until the FF gets `removed`.
Could you please provide more insight into why you believe the files related to the 'old' method should be deleted right away? I'll happily make the changes if I've misunderstood, but I'd appreciate a bit more explanation on this point. 
Thank you!

tomsonpl commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/crowdstrike/crowdstrike_agent_status_client.ts:
```diff
@@ -0,0 +1,191 @@
+  async getAgentStatusFromConnectorAction(agentIds: string[]) {
+    const connectorActionsClient = this.options.connectorActionsClient;
+
+    let connectorList: ConnectorWithExtraFindData[] = [];
+
+    try {
+      connectorList = await connectorActionsClient.getAll();
+    } catch (err) {
+      throw new CustomHttpRequestError(
+        `Unable to retrieve list of stack connectors: ${err.message}`,
+        // failure here is likely due to Authz, but because we don't have a good way to determine that,
+        // the `statusCode` below is set to `400` instead of `401`.
+        400,
+        err
+      );
+    }
+    const connector = connectorList.find(({ actionTypeId, isDeprecated, isMissingSecrets }) => {
+      return actionTypeId === CROWDSTRIKE_CONNECTOR_ID && !isDeprecated && !isMissingSecrets;
+    });
+
+    if (!connector) {
+      throw new CustomHttpRequestError(`No Crowdstrike stack connector found`, 400, connectorList);
+    }
```
Thank you, I totally forgot about `NormalizedExternalConnectorClient` - will change the code 👍

tomsonpl commented on x-pack/plugins/security_solution/public/detections/components/host_isolation/use_host_isolation_action.tsx:
```diff
@@ -159,34 +176,31 @@ export const useHostIsolationAction = ({
 
-      const { isUninstalled, isPendingUninstall } =
-        sentinelOneAgentStatus as AgentStatusInfo[string];
+      const { isUninstalled, isPendingUninstall } = externalAgentStatus as AgentStatusInfo[string];
```
I just renamed the const, not sure about the casting. Would you like me to investigate this? :)

tomsonpl commented on x-pack/plugins/security_solution/public/management/hooks/agents/use_get_agent_status.tsx:
```diff
@@ -21,10 +21,10 @@ interface ErrorType {
-// TODO: 8.15: Remove `useGetSentinelOneAgentStatus` function when `agentStatusClientEnabled` is enabled/removed
-export const useGetSentinelOneAgentStatus = (
+// TODO: 8.15: Remove `useGetExternalAgentStatus` function when `agentStatusClientEnabled` is removed
+export const useGetExternalAgentStatus = (
```
Thanks! Answered in general comment below :)

tomsonpl commented on x-pack/plugins/security_solution/server/endpoint/routes/agent/agent_status_handler.ts:
```diff
@@ -79,17 +82,31 @@ export const getAgentStatusRouteHandler = (
           connectorActionsClient: (await context.actions).getActionsClient(),
         });
+      }
+      if (agentType === 'crowdstrike') {
+        return getCrowdstrikeAgentStatus({
+          agentType,
+          agentIds,
+          logger,
+          connectorActionsClient: (await context.actions).getActionsClient(),
+        });
+      }
```
It is, there are 2 ways supported.

tomsonpl committed:
NormalizedExternalConnectorClient

tomsonpl commented:
@PhilippeOberti thank you for your review 👍  I am gonna improve the testing 👍

tomsonpl committed:
Merge branch 'crowdstrike-agent-status' of github.com:tomsonpl/kibana into crowdstrike-agent-status

tomsonpl commented on x-pack/plugins/security_solution/public/common/components/agents/agent_status/agent_status.tsx:
```diff
@@ -71,19 +74,24 @@ export const AgentStatus = React.memo(
+            {getAgentStatusText(agentStatus.status)}
+          </EuiBadge>
+        ) : (
+          <EuiBadge
+            color={HOST_STATUS_TO_BADGE_COLOR[HostStatus.UNENROLLED]}
+            className="eui-textTruncate"
+            data-test-subj={getTestId('agentStatus')}
+          >
+            {getAgentStatusText(HostStatus.UNKNOWN)}
+          </EuiBadge>
```
What should be stated in the tooltip? isn't the badge enough?

paul-tavares replied:
I thought we had a separate ticket to track the "unknown" changes to the status? (@ashokaditya ??).

The tooltip should provide information to the user as to why we are not able to determine the status of the agent

tomsonpl replied:
Sorry I wasn't aware we discussed this yesterday on grooming, I must have missed that 👍 I'll revert the change

tomsonpl commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/crowdstrike/agent_status.ts:
```diff
@@ -0,0 +1,142 @@
+            ? HostStatus.OFFLINE
+            : HostStatus.UNENROLLED,
+        pendingActions: {
+          isolate:
+            thisAgentDetails.status === CROWDSTRIKE_NETWORK_STATUS.CONTAINMENT_PENDING ? 1 : 0,
+          unisolate:
+            thisAgentDetails.status === CROWDSTRIKE_NETWORK_STATUS.LIFT_CONTAINMENT_PENDING ? 1 : 0,
```
This is the old way indeed, the very same way as `sentinel_one` does it. The getPendingActionsSummary() is used in crowdstrike_agent_status_client. Do you think I should use it here as well?

paul-tavares replied:
Agreed. the Crowdstrike functionality and any other new response actions functionality should not use the "old" way of getting agent status. Only use the new way

tomsonpl committed:
highlighted_fields

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

paul-tavares reviewed (commented):
Ok. Finished up reviewing the rest of the changes - only one other suggestion for the connector changes


_________

re: 

> Before we proceed, I think it's important to align on the approach for fetching agent status, specifically regarding the old and new methods. My understanding is that the new method is activated when the agentStatusClientEnabled FF is enabled. Since this feature flag can be toggled by the user through configuration at any time, I believe we should continue supporting both methods for now until the FF gets removed.
> Could you please provide more insight into why you believe the files related to the 'old' method should be deleted right away? I'll happily make the changes if I've misunderstood, but I'd appreciate a bit more explanation on this point.


We are moving to use the new way of determining agent status starting with 8.15. the new way will be used for all agent types and is not something that is in Tech Preview (as is the case with the Bi-Directional response actions). The crowdstrike functionality is new and still being developed, so I don't see the need to implement the "old" way of fetching agent statuses for crowdstrike as it will likely never be used for crowdstrike - it would just be adding additional tech debt.  So the FF that is controlling the Agent Status work will essentially be a pre-requisite for the Crowdstrike functionality. Having these "prerequisites" is not new - the current SentinelOne functionality is almost all dependent on prior feature flags.

You feel strongly about keeping support for the Old way just let me (us) know and mark the comments as resolved.  Just make sure that you test for both paths for getting the status

tomsonpl committed:
fix test

tomsonpl committed:
Merge branch 'crowdstrike-agent-status' of github.com:tomsonpl/kibana into crowdstrike-agent-status

tomsonpl commented:
Thank you @paul-tavares , I would prefer to stick to supporting two approaches for now, especially because it's gonna provide same support as SentinelOne does. Then we could get rid of Crowdstrike and SentinelOne and whatever else is redundant at the same time (I prefer when the agentStatusClient FF is removed).

ashokaditya commented on x-pack/plugins/security_solution/public/management/components/endpoint_responder/components/offline_callout.tsx:
```diff
@@ -32,16 +33,20 @@ export const OfflineCallout = memo<OfflineCalloutProps>(({ agentType, endpointId
-    enabled: sentinelOneManualHostActionsEnabled && isSentinelOneAgent,
+    enabled:
+      (sentinelOneManualHostActionsEnabled && isSentinelOneAgent) ||
+      (crowdstrikeManualHostActionsEnabled && isSentinelOneAgent),
```
```suggestion
      (crowdstrikeManualHostActionsEnabled && isCrowdStrikeAgent),
```

ashokaditya replied:
This should also be enabled when it is `endpoint` agent and `agentStatusClientEnabled` is `true`.

tomsonpl replied:
Good catch 👍  thanks!

ashokaditya commented on x-pack/plugins/security_solution/public/management/components/endpoint_responder/components/offline_callout.tsx:
```diff
@@ -32,16 +33,20 @@ export const OfflineCallout = memo<OfflineCalloutProps>(({ agentType, endpointId
+      (sentinelOneManualHostActionsEnabled && isSentinelOneAgent) ||
+      (crowdstrikeManualHostActionsEnabled && isSentinelOneAgent),
   });
-
```
Oh! I missed updating this component looks like. 😓  

`endpointDetails` should be fetched only if `agentStatusClientEnabled` is `false`. When it is `true` it should use `data` fetched from `getAgentStatus` to check for agent status.

tomsonpl commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/crowdstrike/crowdstrike_agent_status_client.ts:
```diff
@@ -0,0 +1,191 @@
+
+    return keyBy(agentStatusResponse.data?.resources, 'id');
+  }
+  async getAgentStatuses(agentIds: string[]): Promise<AgentStatusRecords> {
```
I made the other one private, thanks!

tomsonpl committed:
revert crowdstrike_agent_status

tomsonpl committed:
fix

tomsonpl committed:
revert

tomsonpl commented:
@paul-tavares @ashokaditya I reverted all the changes regarding the 'old' way to Sentinel_One agent status, and removed the crowdstrike ones. Could you take another look please? :)

tomsonpl committed:
fix

paul-tavares commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/crowdstrike/crowdstrike_agent_status_client.ts:
```diff
@@ -0,0 +1,171 @@
+/*
```
Can you add tests for this new sub-class

paul-tavares commented on x-pack/plugins/security_solution/public/management/components/endpoint_responder/components/offline_callout.test.tsx:
```diff
@@ -60,6 +60,7 @@ describe('Responder offline callout', () => {
       jest.clearAllMocks();
     });
 
+    // TODO TC: include crowdstrike status test
```
Are you still planning on adding tests for Crowdstrike in this PR?

tomsonpl replied:
Thanks for catching this - no I am not planning to add a case for CS here 👍

tomsonpl replied:
https://github.com/elastic/kibana/pull/180200/commits/5514c114ff92106f8bdcef48913360572cc106b9

paul-tavares replied:
Ok. I'm assuming that it will be done then in the next PR?  _(Just a reminder that we as a general rule want to include test coverage for the code we commit to `main`)_

Question: is this (the display of the offline callout) supposed to be working as expected? 

Looking at the code for the component, I'm not seeing how it is being displayed and I'm having a bit of trouble testing this out manually, so perhaps you can provide a screen capture of this callout being displayed in the UI for Crowdstrike hosts that are offline?

tomsonpl committed:
Merge branch 'main' into crowdstrike-agent-status

ashokaditya commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/lib/base_agent_status_client.ts:
```diff
@@ -17,6 +18,7 @@ export interface AgentStatusClientOptions {
   endpointService: EndpointAppContextService;
   esClient: ElasticsearchClient;
   soClient: SavedObjectsClientContract;
+  connectorActionsClient: ActionsClient;
```
```suggestion
  connectorActionsClient: ActionsClient | undefined;
```

or 

```suggestion
  connectorActionsClient?: ActionsClient;
```

tomsonpl replied:
I think it's required for SentinelOne https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/endpoint/routes/agent/agent_status_handler.ts#L91

ashokaditya replied:
The code you're pointing to does not use this type. That is the old sentinel one agent status method within `x-pack/plugins/security_solution/server/endpoint/services/agent/agent_status.ts`. This type is used by the new agent status client and starting with your PR `connectorActionsClient` is required only for `crowdstrike`

tomsonpl committed:
crowdstrike_agent_status_client test

ashokaditya commented on x-pack/plugins/security_solution/server/endpoint/routes/agent/agent_status_handler.ts:
```diff
@@ -79,12 +81,13 @@ export const getAgentStatusRouteHandler = (
       esClient,
       soClient,
       endpointService: endpointContext.service,
+      connectorActionsClient: (await context.actions).getActionsClient(),
```
If you make the `connectorActionsClient` as optional or that it takes `undefined` (see my later comment) as we do not need it for SentinelOne agent status

```suggestion
      connectorActionsClient:
        agentType === 'crowdstrike' ? (await context.actions).getActionsClient() : undefined,
```

ashokaditya commented on x-pack/plugins/security_solution/server/endpoint/routes/agent/agent_status_handler.ts:
```diff
@@ -79,12 +81,13 @@ export const getAgentStatusRouteHandler = (
-    const getAgentStatusPromise = endpointContext.experimentalFeatures.agentStatusClientEnabled
-      ? agentStatusClient.getAgentStatuses(agentIds)
-      : getAgentStatus({
+    const getAgentStatus = endpointContext.experimentalFeatures.agentStatusClientEnabled
```
```suggestion
    const data = endpointContext.experimentalFeatures.agentStatusClientEnabled
```

ashokaditya commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/crowdstrike/crowdstrike_agent_status_client.ts:
```diff
@@ -0,0 +1,171 @@
+        return acc;
+      }, {});
+
+      const agentStatuses = await this.getAgentStatusFromConnectorAction(agentIds);
```
Since this may fail independently of the rest of the agent info, we should do error handling here. That way we can still show partial agent info instead of showing no status when that happens.

tomsonpl replied:
I think we shouldn't show one without the other, meaning we should throw error - showing just part of the data may be confusing or even worse, may show incorrect state. What do you think?

ashokaditya commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/crowdstrike/crowdstrike_agent_status_client.ts:
```diff
@@ -0,0 +1,171 @@
+            agentStatus?.state === CROWDSTRIKE_STATUS_RESPONSE.ONLINE
+              ? HostStatus.HEALTHY
+              : // TODO TC: not sure what the UNKNOWN is - still to be figured
+              agentStatus?.state === CROWDSTRIKE_STATUS_RESPONSE.UNKNOWN
+              ? HostStatus.UNENROLLED
```
If we don't know anything about the agent state at this point, then just assuming it is unenrolled doesn't seem right.  I would suggest you leave the fallback option as `offline` if we don't have an `Unknown` state.

ashokaditya commented on x-pack/plugins/security_solution/server/endpoint/routes/agent/agent_status_handler.ts:
```diff
@@ -98,7 +101,7 @@ export const getAgentStatusRouteHandler = (
       return response.ok({
         body: {
-          data: await getAgentStatusPromise,
+          data: getAgentStatus,
```
```suggestion
          data,
```

ashokaditya reviewed (changes requested):
Thanks for the changes @tomsonpl . Really appreciate your patience with this. I've a few more changes and suggestions. I strongly feel about: 

1. connectorActionsClient being optional to the agent status client
2. that we should not assume agent is unenrolled if there isn't an "unknown" status yet.
3. that we show partial agent status if connector action fails while computing agent status.

Every other suggestion is either minor naming suggestions or nits. I'll leave them to your discretion.

tomsonpl committed:
get_alert_summary_rows tests

tomsonpl committed:
Merge branch 'crowdstrike-agent-status' of github.com:tomsonpl/kibana into crowdstrike-agent-status

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

tomsonpl committed:
pr nits

tomsonpl committed:
Merge branch 'crowdstrike-agent-status' of github.com:tomsonpl/kibana into crowdstrike-agent-status

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

tomsonpl committed:
loosen the schema

tomsonpl commented on x-pack/plugins/stack_connectors/server/connector_types/crowdstrike/crowdstrike.ts:
```diff
@@ -122,6 +132,20 @@ export class CrowdstrikeConnector extends SubActionConnector<
+        ids: payload.ids,
+      },
+      paramsSerializer,
+      responseSchema: CrowdstrikeGetAgentOnlineStatusResponseSchema,
```
Relaxed a few schemas #b7bb59daa189f04507a6db9db2a5defcde719970

tomsonpl replied:
Ok now I see it makes no sense, I'll try again differently ;p

tomsonpl replied:
https://github.com/elastic/kibana/commit/5514c114ff92106f8bdcef48913360572cc106b9

tomsonpl committed:
Revert "loosen the schema"

This reverts commit b7bb59daa189f04507a6db9db2a5defcde719970.

tomsonpl committed:
schema change

tomsonpl committed:
fix id

tomsonpl committed:
Merge branch 'main' into crowdstrike-agent-status

tomsonpl committed:
Merge branch 'main' into crowdstrike-agent-status

ashokaditya reviewed (approved):
Thanks again for the changes and sitting with me for the live test. Feel free to merge after fixing the API bug we noticed while testing. I've a few suggestions that I'll leave to your discretion.

tomsonpl commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/crowdstrike/crowdstrike_agent_status_client.ts:
```diff
@@ -0,0 +1,172 @@
+            agentStatus?.state === CROWDSTRIKE_STATUS_RESPONSE.ONLINE
+              ? HostStatus.HEALTHY
+              : // TODO TC: not sure what the UNKNOWN is - still to be figured
+              agentStatus?.state === CROWDSTRIKE_STATUS_RESPONSE.UNKNOWN
```
I think that if we remove it, we'll never come back to supporting an `unknown`from CS. I'd prefer to leave it as it is, and get back to it when we have some more data. What do you say?

tomsonpl replied:
Regarding the one below (not sure why I can't answer to that one directly). Are you saying we should silently fail, and just show data as if it succeded?

tomsonpl committed:
check for agent_type

paul-tavares commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/crowdstrike/types.ts:
```diff
@@ -0,0 +1,22 @@
+      last_seen: string;
+      hostname: string;
+    };
+    reduced_functionality_mode: 'no'; // or else?
+    status: string;
+  };
+  cid: string;
+  device: {
+    id: string;
+  };
```
Can you revisit this type definition and sync with the Crowdstrike integration. A few properties here do not look correct to me.  Several property are not defined in the correct structure.

paul-tavares commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/crowdstrike/crowdstrike_agent_status_client.test.ts:
```diff
@@ -0,0 +1,210 @@
+    });
+    it('should get agent statuses', async () => {
+      const agentIds = ['agent1'];
+      const searchResponse = {
```
Being that I think you defined the type incorrectly, can you make sure you add a `type` this variable (and other variables in this PR) so that if you happen to define you mock incorrectly here, at least you will get a type error to signal that it is incorrect.

paul-tavares replied:
Also - once the mock is fixed to the correct structure - can you ensure you have tests to validate each of the expected values `isolated`, `status` and `found` properties since there is logic that calculates it? 

Here are some ideas:

- Test `isolated` is set correctly for the different values that can appear in the Crowdstrike `status` field
- Test that `found` is correctly set to `false`
- Test the different types of agent `status`'s that the crowdstrike agent can have and validate that we calculate our correct "status" value based on it

tomsonpl replied:
Thanks Paul, I adjusted the types, and added some tests.

paul-tavares reviewed (changes requested):
Manually tested the statuses and have found some issues. See below and comments left in the latest code in the PR.


My Tests:

### After `isolate` is sent from Kibana - host shows `isolating` 👍 

<img width="772" alt="image" src="https://github.com/elastic/kibana/assets/56442535/0053a202-04ce-4395-9a49-3f49538feaf0">




### Shows `offline` 👍 

<img width="745" alt="image" src="https://github.com/elastic/kibana/assets/56442535/4556cb29-88c0-4866-b6af-0e98a3f47557">



### Shows `Isolated` if the host status is `contained` ❌ 

This does not seem to be working. 
If the host has no pending actions, and the status of it _(from the crowstrike host index document - field `crowstrike.host.status`) has a value of `contained`, the status should show `isolated` - but its not 😞 

I also think that if the Crowdstrike host has a status of `lift_containment_pending`, we should also show the agent status as `Isolated` _(or maybe `releasing` - but not sure if we have a way to do that without actually returning back `pending_actions.release: 1` which I'm not sure we want to do)_. 


## Some other things

1.  ❌  The `isolate` and `release` flyout does not display `Technical preview` in the flyout title for Crowdstrike alerts <br><img width="771" alt="image" src="https://github.com/elastic/kibana/assets/56442535/81ad2add-b625-4829-a8ff-7ffacb55ec70">

ashokaditya commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/crowdstrike/crowdstrike_agent_status_client.ts:
```diff
@@ -0,0 +1,172 @@
+          agentType: this.agentType,
+          // TODO: check if we should use crowdstrike.cid instead
+          found: agentInfo?.host.id === agentId,
+          isolated: agentInfo?.status === CROWDSTRIKE_NETWORK_STATUS.CONTAINED,
```
Yeah we do this on the frontend using the pendingActions and isolated status. It's [the same logic as we use for endpoint](https://github.com/ashokaditya/kibana/blob/6ea9eb1079cebd136ae1ca1a46fe741941610192/x-pack/plugins/security_solution/public/common/components/agents/agent_status/agent_response_action_status.tsx#L27) and [AgentStatus](https://github.com/elastic/kibana/blob/14d459f5816ba1531f284a19858967809d853dd8/x-pack/plugins/security_solution/public/common/components/agents/agent_status/agent_status.tsx#L89) uses it.

paul-tavares replied:
Right. but for external agents we may need to set the `isolated` property differently, since someone could have isolated the host via the 3rd party system and not from Kibana

ashokaditya replied:
Oh indeed. We need to consider that situation.

tomsonpl replied:
Agree 👍

PhilippeOberti reviewed (approved):
thanks for adding the doc and tests! LGTM for the Threat. Hunting Investigations team

tomsonpl committed:
fix missing isolated label bug

tomsonpl committed:
get agent client test

tomsonpl commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/get_agent_status_client.ts:
```diff
@@ -27,8 +28,8 @@ export const getAgentStatusClient = (
-    // case 'crowdstrike':
-    //   // TODO TC: Currently we do not support CrowdStrike agent status - will be done in a follow up PR
+    case 'crowdstrike':
+      return new CrowdstrikeAgentStatusClient(constructorOptions);
```
There was no current test file, but I created one :) Thanks for catching this missing part 👍

tomsonpl commented on x-pack/plugins/security_solution/server/endpoint/services/agent/clients/crowdstrike/crowdstrike_agent_status_client.test.ts:
```diff
@@ -0,0 +1,210 @@
+                              id: 'agent2',
+                              last_seen: '2023-01-01',
+                            },
+                            status: CROWDSTRIKE_NETWORK_STATUS.CONTAINED,
```
Good catch, thank you! it must be overlooked due to differences between how integration ingests host data, and how the api returned it. 
Fixed: Thanks!

tomsonpl committed:
check releasing status and use const

ashokaditya committed:
mandatory `agentType` for alert flyout actions

fixes the issue where metadata API is being called for non-endpoint agents. With an optional agentType parameter we were defaulting it to `endpoint`. This commit fixes that.

ashokaditya commented on x-pack/plugins/security_solution/public/detections/components/endpoint_responder/use_responder_action_data.ts:
```diff
@@ -160,10 +160,10 @@ export const useResponderActionData = ({
       });
     }
-    if (hostInfo) {
+    if (isEndpointHost && hostInfo) {
```
`isEndpointHost` check is redundant but narrows the `agentType` to `endpoint` so good to have here.

ashokaditya commented on x-pack/plugins/security_solution/public/detections/components/endpoint_responder/use_responder_action_item.tsx:
```diff
@@ -32,15 +33,20 @@ export const useResponderActionItem = (
+    if (isAlertFromSentinelOneEvent({ data: eventDetailsData })) {
+      return 'sentinel_one';
+    }
+    if (isAlertFromCrowdstrikeEvent({ data: eventDetailsData })) {
```
We needed this check here to not default to `endpoint` when the alert detail is for a crowdstrike alert.

tomsonpl committed:
add types

tomsonpl committed:
Merge remote-tracking branch 'origin/crowdstrike-agent-status' into crowdstrike-agent-status

ashokaditya committed:
fix types

ref 7edcf447d65bd5214718afdf860f22ca6d2d5b47

ashokaditya committed:
use sentinel one hook for sentinel_one only

ashokaditya committed:
fix offline callout

tomsonpl committed:
tests

tomsonpl committed:
Merge remote-tracking branch 'origin/crowdstrike-agent-status' into crowdstrike-agent-status

tomsonpl commented:
/ci

ashokaditya committed:
Merge branch 'crowdstrike-agent-status' of github.com:tomsonpl/kibana into crowdstrike-agent-status

tomsonpl commented on x-pack/plugins/security_solution/public/detections/components/endpoint_responder/use_responder_action_item.tsx:
```diff
@@ -32,15 +33,20 @@ export const useResponderActionItem = (
+      return 'sentinel_one';
+    }
+    if (isAlertFromCrowdstrikeEvent({ data: eventDetailsData })) {
+      return 'crowdstrike';
```
Good point, I'll add some tests :)

tomsonpl replied:
Added

tomsonpl committed:
add simple tests

tomsonpl committed:
lint and revert ff

tomsonpl committed:
Merge remote-tracking branch 'origin/crowdstrike-agent-status' into crowdstrike-agent-status

tomsonpl committed:
rename

tomsonpl committed:
Merge branch 'main' into crowdstrike-agent-status

paul-tavares reviewed (approved):
Thank you for all the updates 🙏

tomsonpl commented:
Thank you everyone for the help 🙇
